sdk/servicebus/Azure.Messaging.ServiceBus/tests/Processor/ProcessorLiveTests.cs

```diff
@@ -513,6 +514,45 @@ Task ProcessMessage(ProcessMessageEventArgs args)
             }
         }
 
+        [Test]
+        public async Task StopProcessingAdheresToTokenSLA()
+        {
+            await using (var scope = await ServiceBusScope.CreateWithQueue(
+                enablePartitioning: false,
+                enableSession: false))
+            {
+                // very long timeout
+                await using var client = CreateClient(tryTimeout: 60);
+                var sender = client.CreateSender(scope.QueueName);
+                await sender.SendMessageAsync(GetMessage());
+                await using var processor = client.CreateProcessor(scope.QueueName, new ServiceBusProcessorOptions
+                {
+                    AutoCompleteMessages = true,
+                });
+                var tcs = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+                Task ProcessMessage(ProcessMessageEventArgs args)
+                {
+                    tcs.TrySetResult(true);
+                    return Task.CompletedTask;
+                }
+                processor.ProcessMessageAsync += ProcessMessage;
+                processor.ProcessErrorAsync += ExceptionHandler;
+
+                await processor.StartProcessingAsync();
+                await tcs.Task;
+                await Task.Delay(2000); // better way to do this?
+
+                using var cancellationTokenSource = new CancellationTokenSource(TimeSpan.FromSeconds(3));
+
+                var start = DateTime.UtcNow;
+                await processor.StopProcessingAsync(cancellationTokenSource.Token);
+                var stop = DateTime.UtcNow;
+
+                Assert.That(stop - start,  Is.EqualTo(TimeSpan.FromSeconds(3)).Within(TimeSpan.FromSeconds(3)));
```

Comment:
Hmm, does this test pass? The issue with stopping the processor is that we await the receive call that the receiver managers are doing. So when you call StopProcessing, it can still take up to 60s (the TryTimeout default value from the RetryOptions) for it to actually stop. The behavior we want is that when a user calls StopProcessing (or Close/Dispose) any ongoing Receive operations are immediately cancelled, but any ongoing user handlers that are running are allowed to complete.

Reply:
![image](https://user-images.githubusercontent.com/174258/112798218-e5ba6d80-906c-11eb-96a1-1eb0b67f0258.png)

> In a recent conversation with the AMQP library team, it was suggested that force-closing an active AMQP link could be safely used to cancel an operation in-flight. This would potentially allow the client types to respect cancellation while also gracefully shutting down and cleaning up.

This is what this code does.

The flow is the following:

- Once `StopProcessingAsync` is called the cancellation token that signals the event handlers things are about to shutdown is triggered (as before).
- This leaves handlers the chance to complete either indefinitely (when no cancellation token provided to `StopProcessingAsync`) or up "SLA" enforced by the token passed into the `StopProcessingAsync`. Once the token triggers the link is forcefully closed which then makes the `Task.Factory.FromAsync` part return and throw right next due to the token being canceled.

Reply:
I may be overlooking something, but the gap that I see here is that the token that you're passing into `StopProcessingAsync` is only used in the scope of that method to request that the stop operation be aborted.   It doesn't get propagated to the handlers, so it really only allows for us to abort when the semaphore is taking too long to acquire.   No?

Reply:
With the cooperative cancellation approach of .NET the cancellation token only applied to the flow of the method you pass it in. So from the perspective of the caller it really means stop processing and here is the SLA that I'm going to give you to do so. And you are correct that the token is not "lifted" to the handlers. The token in the handlers is immediately flagged to give the handler methods a possibility to stop doing what they are doing to gracefully stop things (and yes graceful can mean throwing because of cancellation). Once the SLA of the stop is reached the stop processing method returns and from that stand point it is like switching the lights off. 

>  to the handlers, so it really only allows for us to abort when the semaphore is taking too long to acquire. No?

Not sure what you mean here. Which semaphore do you mean? There are multiple semaphores at play. Once is to make sure a single caller can stop and another one is to make sure the concurrency settings are guaranteed.

Reply:
Apologies for the confusion.   The point that I was making is that cancellation in `StopProcessingAsync` is best effort.   We intentionally don't honor it if it would result in corruption or inconsistent state.   If you make it past the check on Line 626, that's the point of no return.   In the case that the processing task takes too long to stop, it will and should) ignore the token passed in.

Reply:
> Not sure what you mean here. Which semaphore do you mean? 

I'm specifically referring to the `_processingStartStopSemaphore` call on line 620.   That's very nearly the last point that we respect cancellation from the token passed in and the only thing that I'd expect to have variable timing before we pass the point of no return.

Reply:
> The flow is the following:
Once StopProcessingAsync is called the cancellation token that signals the event handlers things are about to shutdown is triggered (as before).
This leaves handlers the chance to complete either indefinitely (when no cancellation token provided to StopProcessingAsync) or up "SLA" enforced by the token passed into the StopProcessingAsync. Once the token triggers the link is forcefully closed which then makes the Task.Factory.FromAsync part return and throw right next due to the token being canceled.

Ah I see. It is an interesting approach. Basically users would have to specify their stop SLA via the StopProcessingAsync call. The alternative that I was considering is that we do this automatically by cancelling in-flight receive calls, but still allowing ongoing event handlers to finish. This would be how it would work if the underlying AMQP library had cancellation token support in the Receive API.

Reply:
In case you are interested in exploring this other approach, here is the [code](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ReceiverManager.cs#L83) that would cause StopProcessing to take up to 60s in the default case.

Reply:
Fair enough. I followed the proposal in the original issue description in this spike. Maybe I find a few minutes to do the other approach as a community PR.

Reply:
@JoshLove-msft @jsquire Quick question. I think I have a good proposal. Just wanted to clarify something. My plan is to implement it in the AMQP abstractions of the service bus library as close to the bare metal as possible so that regular receives can also properly benefit from the cancellation token and not just the processor. What do you think?

Reply:
I think this makes sense. Ideally it would be in the AMQP library itself.

Reply:
I strongly agree with both of those statements.    I don't see much opportunity for this to move into the AMQP library, given the history of attempts there.... the next best thing would be to go with Daniel's proposal.   Assuming that the "close it and things don't blow up" works, the form that is used here is quite solid.

Reply:
Here we go gents https://github.com/Azure/azure-sdk-for-net/pull/19955